### PR TITLE
Fix cases when AI ignores some objects during path evaluation

### PR DESCRIPTION
**Pull request by ihhub**
std::binary_search works only if an array is partially sorted in respect to value - https://en.cppreference.com/w/cpp/algorithm/binary_search

~~Since **getObjectsOnTheWay()** method returns unsorted array we have to sort it in order to fulfill binary search requirements.~~
**_mapObjects** wasn't been completely sorted during fog revealing events.

### CODE REVIEW

File: src/fheroes2/ai/normal/ai_normal_hero.cpp [hunk 1]
```diff
@@ -1673,6 +1673,8 @@ namespace AI
             if ( !isDimensionDoor ) {
                 // Dimension door path does not include any objects on the way.
                 std::vector<IndexObject> list = _pathfinder.getObjectsOnTheWay( destination );
+                std::sort( list.begin(), list.end() );
```
Comment (oleg-derevenetz) [inline src/fheroes2/ai/normal/ai_normal_hero.cpp:1676]: Hi @ihhub but there is `_mapObjects` should be sorted, and not the `list`.

Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_hero.cpp:1676]: Woops... I am completed blind.

Comment (ihhub) [reply src/fheroes2/ai/normal/ai_normal_hero.cpp:1676]: Boom! Even better. **_mapObjects** can be unsorted during fog revealing.

Comment (oleg-derevenetz) [reply src/fheroes2/ai/normal/ai_normal_hero.cpp:1676]: > Boom! Even better. _mapObjects can be unsorted during fog revealing.

Yes.

**Review by oleg-derevenetz — CHANGES_REQUESTED** [review]
Hi @ihhub I recommend also to add the assertion that `_mapObjects` is `std::is_sorted` somewhere near that `std::binary_search` (maybe only for the debug build?) and also to add the comment why we are using the `upper_bound` here.

**Review by oleg-derevenetz — APPROVED** [review]

**Review by Districh-ru — APPROVED** [review]
I tested AI on two XL maps, one time with debug hero (with Dimension Door spell) for a half an hour and found no bugs. Also I did not notice strange AI behavior.

**Pull request merged**